### PR TITLE
feat: auto generate components.json, global plugin for tests and vetur tags

### PR DESCRIPTION
**Pull request by pi0**
This PR allows optional integration with other tools and frameworks.

Vetur integration PR: vuejs/vetur/pull/1921 

<details>
  <summary>.nuxt/components/components.js</summary>

```js
module.exports = function install(Vue) {
// components/base/Button.vue
Vue.component('BaseButton', require('../test/fixture/components/base/Button.vue').default)

// components/base/Button.vue
Vue.component('LazyBaseButton', Promise.resolve(require('../test/fixture/components/base/Button.vue').default))

// components/icons/Home.vue
Vue.component('IconHome', require('../test/fixture/components/icons/Home.vue').default)

// components/icons/Home.vue
Vue.component('LazyIconHome', Promise.resolve(require('../test/fixture/components/icons/Home.vue').default))

// components/Bar.ts
Vue.component('Bar', require('../test/fixture/components/Bar.ts').default)

// components/Bar.ts
Vue.component('LazyBar', Promise.resolve(require('../test/fixture/components/Bar.ts').default))

// components/Foo.vue
Vue.component('Foo', require('../test/fixture/components/Foo.vue').default)

// components/Foo.vue
Vue.component('LazyFoo', Promise.resolve(require('../test/fixture/components/Foo.vue').default))
}
```
</details>

<details>
  <summary>.nuxt/vetur/tags.json</summary>

```json
{
  "BaseButton": {
    "description": "Auto imported from components/base/Button.vue"
  },
  "LazyBaseButton": {
    "description": "Auto imported from components/base/Button.vue"
  },
  "IconHome": {
    "description": "Auto imported from components/icons/Home.vue"
  },
  "LazyIconHome": {
    "description": "Auto imported from components/icons/Home.vue"
  },
  "Bar": {
    "description": "Auto imported from components/Bar.ts"
  },
  "LazyBar": {
    "description": "Auto imported from components/Bar.ts"
  },
  "Foo": {
    "description": "Auto imported from components/Foo.vue"
  },
  "LazyFoo": {
    "description": "Auto imported from components/Foo.vue"
  }
}
```
</details>

<details>
  <summary>.nuxt/components/components.json</summary>

```json
[
  {
    "pascalName": "BaseButton",
    "kebabName": "base-button",
    "shortPath": "components/base/Button.vue"
  },
  {
    "pascalName": "LazyBaseButton",
    "kebabName": "lazy-base-button",
    "shortPath": "components/base/Button.vue",
    "async": true
  },
  {
    "pascalName": "IconHome",
    "kebabName": "icon-home",
    "shortPath": "components/icons/Home.vue"
  },
  {
    "pascalName": "LazyIconHome",
    "kebabName": "lazy-icon-home",
    "shortPath": "components/icons/Home.vue",
    "async": true
  },
  {
    "pascalName": "Bar",
    "kebabName": "bar",
    "shortPath": "components/Bar.ts"
  },
  {
    "pascalName": "LazyBar",
    "kebabName": "lazy-bar",
    "shortPath": "components/Bar.ts",
    "async": true
  },
  {
    "pascalName": "Foo",
    "kebabName": "foo",
    "shortPath": "components/Foo.vue"
  },
  {
    "pascalName": "LazyFoo",
    "kebabName": "lazy-foo",
    "shortPath": "components/Foo.vue",
    "async": true
  }
]
```
</details>

### CODE REVIEW

File: src/scan.ts [hunk 1]
```diff
@@ -47,9 +49,10 @@ export async function scanComponents (dirs: ScanDir[]): Promise<Component[]> {
       const pascalName = pascalCase(fileName)
       const kebabName = kebabCase(fileName)
 
+      const meta = { filePath, pascalName, kebabName }
       components.push(
-        prefixComponent(prefix, { pascalName, kebabName, import: `require('${filePath}').default` }),
-        prefixComponent(LAZY_PREFIX, prefixComponent(prefix, { pascalName, kebabName, import: `function () { return import('${filePath}' /* webpackChunkName: "components/${kebabName}" */) }` }))
+        prefixComponent(prefix, { ...meta, import: `require('${filePath}').default` }),
+        prefixComponent(LAZY_PREFIX, prefixComponent(prefix, { ...meta, async: true, import: `function () { return import('${filePath}' /* webpackChunkName: "components/${kebabName}" */) }` }))
```
Comment (Atinux) [inline src/scan.ts:55]: Why not using `/* webpackChunkName: "${filePath}" */` ?

Comment (Atinux) [reply src/scan.ts:55]: Also removing `srcDir` and I guess the extension?

Comment (pi0) [reply src/scan.ts:55]: Seems good idea. I was already doing this for generated meta json

File: src/index.ts [hunk 2]
```diff
@@ -57,7 +57,7 @@ export default <Module> function () {
 
     this.options.build!.transpile!.push(...componentDirs.filter(dir => dir.transpile).map(dir => dir.path))
 
-    let components = await scanComponents(componentDirs)
+    let components = await scanComponents(componentDirs, this.options.srcDir as string)
```
Comment (kevinmarrec) [inline src/index.ts:60]: `srcDir` type being `string | undefined`, using `!` (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html -> Non-null assertion operator) make it resolved to type `string`, which is cleaner :)

```suggestion
    let components = await scanComponents(componentDirs, this.options.srcDir!)
```

File: src/index.ts [hunk 3]
```diff
@@ -81,7 +81,7 @@ export default <Module> function () {
           return
         }
 
-        components = await scanComponents(componentDirs)
+        components = await scanComponents(componentDirs, this.options.srcDir as string)
```
Comment (kevinmarrec) [inline src/index.ts:84]:
```suggestion
        components = await scanComponents(componentDirs, this.options.srcDir!)
```

File: src/scan.ts [hunk 4]
```diff
@@ -31,28 +33,35 @@ function prefixComponent (prefix: string = '', { pascalName, kebabName, ...rest
...
         continue
       }
+      processedPaths.push(filePath)
 
       const fileName = basename(file, extname(file))
       const pascalName = pascalCase(fileName)
       const kebabName = kebabCase(fileName)
+      const shortPath = filePath.replace(srcDir, '').replace(/\\/g, '/').replace(/^\//, '')
+
+      let cuhnkName = shortPath.replace(extname(shortPath), '')
```
Comment (kevinmarrec) [inline src/scan.ts:54]: You probably meant `chunkName` :P

File: src/index.ts (hunk 2 above)
Comment (pi0) [reply src/index.ts:60]: Didn't know that thanks for the tip 😊 BTW why it is nullable? In nuxt context it always has value

File: src/scan.ts (hunk 4 above)
Comment (pi0) [reply src/scan.ts:54]: It is another language 😂😂😂

File: src/index.ts (hunk 2 above)
Comment (kevinmarrec) [reply src/index.ts]: @pi0 It's cause `options` is of type `NuxtConfiguration`, I once tried to make it Required<NuxtConfiguration> to skip nullables but it applies only to top level properties.

Comment (pi0) [reply src/index.ts]: I will message you re this ;)

**Comment by pi0** [comment]
@kevinmarrec Running tests on one OS obviously results in not complete coverage unless we ignore platform-specific lines for coverage which IMO is hiding truth. [codecov](https://codecov.io/gh/nuxt/components/pull/14/tree/src) shows full coverage (so badge is 100 ;)

**Comment by kevinmarrec** [comment]
@pi0 I see, I think it's cause Windows Github Actions job is runned after Ubuntu one, which means it uploads coverage of Windows as last, which run 100% **statement** coverage, but not full **100%**

I guess we can't have full 100% on both cause of the `os` condition

**Comment by pi0** [comment]
@kevinmarrec Yep :P Combined statement coverage is only 100 which codecov does at the end.

**Comment by designy** [comment]
thanks for this feature, please add support for kebab-case convention in component name autocompletion.

**Comment by pi0** [comment]
@designy PR welcome :)

**Pull request merged**